src/codegen/aggregation.cpp

```diff
@@ -170,82 +159,156 @@ void Aggregation::CreateInitialValues(
-                                        aggregate_info.storage_index);
-        auto delta = Value{type::Type::TypeId::BIGINT, codegen.Const64(1)};
-        next = curr.Add(codegen, delta);
-        break;
-      }
-      case ExpressionType::AGGREGATE_COUNT_STAR: {
-        // COUNT(*) is always stored first with a source equal to INT_MAX. All
-        // other COUNT(*)'s refer to that one.
-        if (source != std::numeric_limits<uint32_t>::max()) {
-          continue;
+    if (aggregate_info.aggregate_type == ExpressionType::AGGREGATE_AVG) {
+      // AVG() aggregates are metadata only
+      continue;
+    }
+
+    const Value &update = next_vals[aggregate_info.source_index];
+
+    bool is_nullable = null_bitmap.IsNullable(aggregate_info.storage_index);
+    bool is_count =
+        aggregate_info.aggregate_type == ExpressionType::AGGREGATE_COUNT ||
+        aggregate_info.aggregate_type == ExpressionType::AGGREGATE_COUNT_STAR;
+    if (is_count || (!is_nullable && !IsGlobal())) {
+      // This aggregate is not nullable, generate fast path
+      DoAdvanceValue(codegen, storage_space, aggregate_info, update);
+    } else {
+      // This aggregate is NULL-able, we need to check if the update value is
+      // NULL, and whether the current value of the aggregate is NULL.
+
+      // There are two cases we handle:
+      // (1). If neither the update value or the current aggregate value are
+      //      NULL, we simple do the regular aggregation without NULL checking.
+      // (2). If the update value is not NULL, but the current aggregate **is**
+      //      NULL, then we just store this update value as if we're creating it
+      //      for the first time.
+      //
+      // If either the update value or the current aggregate are NULL, we have
+      // nothing to do.
+
+      llvm::Value *update_not_null = update.IsNotNull(codegen);
+      llvm::Value *agg_null =
+          null_bitmap.IsNull(codegen, aggregate_info.storage_index);
+
+      llvm::Value *curr_val =
+          null_bitmap.ByteFor(codegen, aggregate_info.storage_index);
+
+      If valid_update{codegen, update_not_null};
+      {
+        If agg_is_null{codegen, agg_null};
+        {
+          // (2)
+          switch (aggregate_info.aggregate_type) {
+            case ExpressionType::AGGREGATE_SUM:
+            case ExpressionType::AGGREGATE_MIN:
+            case ExpressionType::AGGREGATE_MAX: {
+              storage_.SetValue(codegen, storage_space,
+                                aggregate_info.storage_index, update,
+                                null_bitmap);
+              break;
+            }
+            case ExpressionType::AGGREGATE_COUNT: {
+              codegen::Value one{type::BigInt::Instance(), codegen.Const64(1)};
+              storage_.SetValue(codegen, storage_space,
+                                aggregate_info.storage_index, one, null_bitmap);
+              break;
+            }
+            default: { break; }
```

Comment:
COUNT(*) is never NULL-able and will always be handled in `DoAdvanceValue()`. I've simplified the code a little in the latest push to clear this up.